go/controller/cmd/main.go

```diff
@@ -381,3 +390,45 @@ func waitForReady(f func() error, timeout, interval time.Duration) error {
 		time.Sleep(interval)
 	}
 }
+
+// ConfigureNamespaceWatching sets up the controller manager to watch specific namespaces
+// based on the provided configuration. It returns the list of namespaces being watched,
+// or nil if watching all namespaces.
+func ConfigureNamespaceWatching(watchNamespaces string) map[string]cache.Config {
+	watchNamespacesList := filterValidNamespaces(strings.Split(watchNamespaces, ","))
+	if len(watchNamespacesList) == 0 {
+		setupLog.Info("Watching all namespaces (no valid namespaces specified)")
+		return map[string]cache.Config{"": {}}
+
+	}
+	setupLog.Info("Watching specific namespaces at cache level", "namespaces", watchNamespacesList)
+
+	namespacesMap := make(map[string]cache.Config)
+	for _, ns := range watchNamespacesList {
+		namespacesMap[ns] = cache.Config{}
+	}
+
+	return namespacesMap
+}
+
+// filterValidNamespaces removes invalid namespace names from the provided list.
+// A valid namespace must be a valid DNS-1123 label.
+func filterValidNamespaces(namespaces []string) []string {
+	var validNamespaces []string
+
+	for _, ns := range namespaces {
+		if strings.TrimSpace(ns) == "" {
+			continue
+		}
+
+		if errs := validation.IsDNS1123Label(ns); len(errs) > 0 {
+			setupLog.Info("Ignoring invalid namespace name",
+				"namespace", ns,
+				"validation_errors", strings.Join(errs, ", "))
+		} else {
+			validNamespaces = append(validNamespaces, ns)
+		}
+	}
+
+	return validNamespaces
+}
```

Comment:
I'm going to approve, but in a follow-up please move these functions and tests into a different package.